### ISSUE TITLE
Way to reset the chat?

### ISSUE THREAD

**Issue by mkultra333** [kuvaus/LlamaGPTJ-chat#10]
My chats often get stuck in repetitive loops, or there are situations where I want to issue a new command with no memory of the previous prompts and responses.  On GPT4ALL you can press the button to clear the chat to do that.

In LlamaGPTJ-chat I've been shutting it down and restarting to achieve the same, but this comes with the time overhead  and potential memory issues of re-loading the model.  I know it's not much, but shaving 5 seconds off the time by not having to reload would be handy. 

Is there already a way to reset the chat (especially the gptj models since that's what I'm using) and if not, can you suggest how I might do this?  As a wild guess I did a quick experiment that used a modified version of code from where the context gets cleared if it runs out of room but it didn't work, it'd crash or get stuck in loops or something.  So I was looking at this bit of code as a possible hint at how to do it, but don't know if that's really relevant at all.

        // Check if the context has run out...
        if (promptCtx.n_past + batch.size() > promptCtx.n_ctx) {
            const int32_t erasePoint = promptCtx.n_ctx * promptCtx.contextErase;
            // Erase the first percentage of context from the tokens...
            std::cerr << "GPTJ: reached the end of the context window so resizing\n";
            promptCtx.tokens.erase(promptCtx.tokens.begin(), promptCtx.tokens.begin() + erasePoint);
            promptCtx.n_past = promptCtx.tokens.size();
            recalculateContext(promptCtx, recalculateCallback);
            assert(promptCtx.n_past + batch.size() <= promptCtx.n_ctx);
        }

**Comment by kuvaus** [comment]
Thats a good suggestion! I'll look into it. Thanks for the code too! :)

And you're right, theres no reset functionality right now. Even the `recalculate_callback` function in `src/chat.cpp` has no extra stuff:

```c++
    auto recalculate_callback = [](bool is_recalculating) {
        // You can handle recalculation requests here if needed
        return is_recalculating;
    };
```

I need to check how the button in GPT4ALL GUI chat does it and copy that... but I'm not an expert in QT.
I agree that it would be handy so I'll try to put it in the next version if I can figure it out. :)


Some thoughts for starters: There's probably two ways to do it 1) either with the context_erase like you showed (somehow setting it to 100%) or 2) with regenerating the whole prompt_context struct like when its done at the start.

Maybe add something like this in the `get_input()` function in `src/chat.cpp`:

```c++
    if (input == "resetchat") {
	//then either:
	//reset the context here somehow
	//or somehow make a new function and call it from here:
	//reset_context(whatever parameters it needs)
    }
```
Then you could write `resetchat` during chatting and it would work just like the button...

**Comment by mkultra333** [comment]
I didn't write that code, that's the original from the program, where it tokenizes the new user input and checks if the context is full.  I barely have any idea what it does. :)

I tried setting the context to 100%, and other values, but it always either crashed or just hung.  If that is the way to do it then I guess it needs to be done in the right place, I tried in a few places near where the original code was.

The trigger might need to be something like "*resetchat" or "###resetchat" that won't occur in normal prompts and is easy to filter.

**Comment by kuvaus** [comment]
This ended up being a very quick thing to code so I pushed an update already.

In the GPT4ALL GUI chat the button simply makes a new empty context. I ended up doing very similar thing, emptied context (apart from settings).

Didn't do proper testing so write here if you find it did not work. :)

> The trigger might need to be something like "*resetchat" or "###resetchat" that won't occur in normal prompts and is easy to filter.

Yup. I agree. The resetchat trigger only works in the user input function for now. And since its written as a single word, I doubt it  comes in conversations spontaneously. (If the AI ends up using that trigger in response, or if its used in the template, or in a loaded prompt,  it currently should not trigger anything.) I hope.

**Comment by mkultra333** [comment]
Just did a quick test and it seems to work great.  Thanks!

**Comment by kuvaus** [comment]
Hahah. Thanks for the edit!

It never even occurred to me that the user could keep spamming the same command repeatedly. :)
Fixed in v0.2.3. The fix actually made things simpler, which is always nice. (The input function now simply calls itself after reset.)

**Comment by mkultra333** [comment]
Works great.  Thanks.
